File: TodoTest.Web/Scripts/app/controllers/TodoController.js
Todo.controller('TodoController', function ($scope, $log, TodoService) {

    $scope.currentPage = 1;
    $scope.pageSize = 5;
    $scope.todos = [];
    $scope.pagecounter = [];

    $scope.newTodoName = "";
    $scope.pageright = "";
    $scope.pageleft = "";

    $scope.init = function () {
        //var promise = TodoService.get($scope.pageSize, $scope.pageSize);
      
        $scope.then(function (response) {
            $scope.todos = response;
            $scope.todos2 = response;
            $scope.pagecounter = response;
        });
    };

    $scope.paginator = function (page) {
       
        var promise = TodoService.get(page, $scope.pageSize, $scope.newTodoName);
        promise.then(function (response) {
            $scope.todos2 = new Array();

            var s = eval(response);
            var l = JSON.parse(s);

            for (var i = 0; i < l.length; i++) {

                $scope.todos2.push(new TodoViewModel(l[i].Id, l[i].Name, false));
            }

        });
    };

    $scope.addTodo = function () {
        if (!$scope.newTodoName) return;
        
        var promise = TodoService.add($scope.newTodoName);
        promise.then(function (response) {
            $scope.todos.push(new TodoViewModel(response.Id, response.Name, false));
            if (_.size($scope.todos2) > 0)
            {
                TodoService.todolist(response);
                $scope.todos2.push(new TodoViewModel(response.Id, response.Name, false));
            }
  
            $scope.newTodoName = "";
        });
    };

    $scope.inputUpdate = function (ev)
    {
        if (!$scope.newTodoName) return;
        if (ev.which != 13) {
            setTimeout(function () {


                var promise = TodoService.inputUpdate($scope.newTodoName);

                promise.then(function (response) {

                    $scope.todos2 = new Array();

                    var s = eval(response);
                    var l = JSON.parse(s);

                    for (var i = 0; i < l.length; i++) {

                        $scope.todos2.push(new TodoViewModel(l[i].Id, l[i].Name, false));
                    }

                    var count = ($scope.todos2.length / 10) + 1;
                    for (var i = 0; i < count; i++) {
                        $scope.pagecounter.push(i);
                     }
                });
            }, 1000);
        }
        else
        {
            $scope.todos = $scope.todos2;
            TodoService.inputUpdate();
        }


        return _.size($scope.todos2) > 0;
        
    }

    $scope.searchUpdate = function () {
        if (!$scope.newTodoName) return;

        var promise = TodoService.add($scope.newTodoName);
        promise.then(function (response) {
            $scope.todos2.push(new TodoViewModel(response.Id, response.Name, false));
            $scope.newTodoName = "";
        });
     }



    $scope.removeTodo = function (id, removename, name) {
        var promise = TodoService.remove(id, removename, name);
        promise.then(function (response) {
            if (removename == "paginator")
            {
                $scope.todos2 = _.reject($scope.todos2, function (x) { return x.Id === id; });
            }
            else
            {
                $scope.todos = _.reject($scope.todos, function (x) { return x.Id === id; });
            }
            
        });
    };

    $scope.completeTodo = function(todo) {
        var promise = TodoService.update(todo);
        promise.then(function (response) {
            // ui updated by bindings
        });
    };

    $scope.hasAnyTodos = function () {
        return _.size($scope.todos) > 0;
    };

    $scope.fromtodos2totodo = function ()
    {
        $scope.todos = $scope.todos2;
        TodoService.inputUpdate();
    
    }

    function TodoViewModel(id, name, isComplete) {
        this.Id = id;
        this.Name = name;
        this.Completed = isComplete;
    }
});

File: TodoTest.Web/Controllers/TodoController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Web;
using System.Web.Mvc;
using TodoTest.Web.Data;
using TodoTest.Web.Models;
using TodoTest.Web.Models.ViewModels;

namespace TodoTest.Web.Controllers
{
    public class TodoController : Controller
    {
        private readonly IStore store;

        private static readonly log4net.ILog log = log4net.LogManager.GetLogger
    (System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
   

        public TodoController(IStore store)
        {
            this.store = store;
        }
       
        public JsonResult List(int page, int pageSize, string name)
        {
            try
            {
                var todos = MvcApplication.TodoList.Skip((page) * pageSize).Take(pageSize);
                MvcApplication.PaginatorList = new List<Todo>();
                foreach (var item in todos)
                {
                    if (item.Name.ToLower().Contains(name.ToLower()))
                    {
                        MvcApplication.PaginatorList.Add(item);
                    }
                }
                
                System.Web.Script.Serialization.JavaScriptSerializer jSearializer =
      new System.Web.Script.Serialization.JavaScriptSerializer();
                string s = jSearializer.Serialize(MvcApplication.PaginatorList);
                return new JsonResult
                {
                    Data = s,
                    JsonRequestBehavior = JsonRequestBehavior.AllowGet
                };
            }
            catch (Exception ex) 
            {
                log.Debug("Paginator fungerar ej", ex);
            }

            return null;

        }

        [HttpPost]
        public JsonResult Add(string name)
        {
            try
            {
                if (string.IsNullOrEmpty(name))
                {
                    throw new HttpException((int)HttpStatusCode.BadRequest, "name cant be empty");
                }
                
                var createdTodo = store.Add(new Todo(name));
                return new JsonResult
                {
                    Data = createdTodo
                };
            }
            catch (Exception ex) 
            {
                log.Debug("Att lägga nya todos fungerar ej", ex);
            }

            return null;
        }

        [HttpPost]
        public void AddToDoList(TodoViewModel todo) 
        {
            MvcApplication.TodoList.Add(new Todo(todo.Name));
        }

        [HttpPost]
        public JsonResult Remove(string guid, string removename, string name)
        {
            try
            {
                if (MvcApplication.TodoList.Count > 0 && MvcApplication.TodoList != null && removename == "paginator")
                {
                    foreach (var item in MvcApplication.PaginatorList) 
                    {
                        if (item.Id == Guid.Parse(guid)) 
                        {
                            MvcApplication.PaginatorList.Remove(item);
                            MvcApplication.TodoList.Remove(item);
                        }
                    }

                    foreach (var item in MvcApplication.TodoList)
                    {
                        if (item.Id == Guid.Parse(guid))
                            MvcApplication.TodoList.Remove(item);
                    }
                    
                   return new JsonResult
                    {
                        Data = new TodoViewModel(new Todo(name))
                    };
                }
                else
                {
                    var deletedToDo = store.Delete<Todo>(Guid.Parse(guid));
                    if (deletedToDo == null)
                    {
                        throw new HttpException((int)HttpStatusCode.NotFound, "todo not found");
                    }

                    return new JsonResult
                    {
                        Data = new TodoViewModel(deletedToDo)
                    };
                }
            }
            catch (Exception ex)
            {
                log.Debug("Att ta bort todos fungerar ej", ex);
            }

            return null;
        }
        
        [HttpPost]
        public JsonResult Update(TodoViewModel todo)
        {
            var updatedTodo = store.Update(new Todo(todo.Name, todo.Completed) { Id = Guid.Parse(todo.Id) });
            if (updatedTodo == null)
            {
                throw new HttpException((int)HttpStatusCode.NotFound, "todo not found");
            }

            return new JsonResult
            {
                Data = new TodoViewModel(updatedTodo)
            };
        }

        [HttpPost]
        public JsonResult InputUpdate(string name)
        {
            try
            {
                var list = store.Get<Todo>(0, 50);
                MvcApplication.TodoList = new List<Todo>();

                if (list != null && name != null)
                {
                    foreach (var item in list)
                    {
                        if (item.Name.ToLower().Contains(name.ToLower()))
                        {
                            MvcApplication.TodoList.Add(item);
                        }

                    }
                }
                
                System.Web.Script.Serialization.JavaScriptSerializer jSearializer =
                  new System.Web.Script.Serialization.JavaScriptSerializer();
                string s = jSearializer.Serialize(MvcApplication.TodoList);

                return new JsonResult
                {
                    Data = s
                };
            }
            catch (Exception ex) 
            {
                log.Debug("Att söka todos fungerar ej", ex);
            }

            return null;

        }
    }

}
